File: examples/python/send_traces.py
import os

from langfuse import get_client, propagate_attributes

SCORE_NAME = "human-quality"
SCORE_VALUE = 1
SCORE_COMMENT = "Human review: answer is accurate and actionable."
COST_DETAILS = {"input": 0.038, "output": 0.044, "total": 0.082}


def require_langfuse_env():
    missing = [
        key
        for key in ("LANGFUSE_BASE_URL", "LANGFUSE_PUBLIC_KEY", "LANGFUSE_SECRET_KEY")
        if not os.getenv(key)
    ]
    if missing:
        raise RuntimeError(
            f"Missing required Langfuse environment variables: {', '.join(missing)}. "
            "Set LANGFUSE_BASE_URL, LANGFUSE_PUBLIC_KEY, and LANGFUSE_SECRET_KEY before running this example."
        )


def answer_checkout_question(langfuse):
    with langfuse.start_as_current_observation(
        as_type="span",
        name="checkout support trace",
        input={
            "question": "Why was my card declined?",
            "locale": "en-US",
            "cart_total": 128.45,
        },
        metadata={
            "example": "agentpond-python",
            "route": "/support/checkout",
            "tenant": "acme-retail",
        },
    ) as trace:
        with propagate_attributes(
            user_id="user_42",
            session_id="sdk-example-session",
            trace_name="checkout support trace",
            metadata={
                "example": "agentpond-python",
                "feature": "checkout-support",
                "release": "2026-06-15",
            },
        ):
            with langfuse.start_as_current_observation(
                as_type="span",
                name="load customer order context",
                input={"user_id": "user_42", "order_id": "ord_2026_001"},
            ) as lookup:
                lookup.update(
                    output={
                        "payment_status": "declined",
                        "decline_code": "insufficient_funds",
                        "attempts": 1,
                    }
                )

            with langfuse.start_as_current_observation(
                as_type="generation",
                name="draft support response",
                model="gpt-5.5-mini",
                input=[
                    {
                        "role": "system",
                        "content": "Answer checkout support questions with concise next steps.",
                    },
                    {
                        "role": "user",
                        "content": "Why was my card declined?",
                    },
                ],
                metadata={"provider": "example-fixture", "temperature": 0.2},
            ) as generation:
                generation.update(
                    output={
                        "answer": "The bank declined the authorization. Try another card or contact your bank.",
                        "confidence": 0.92,
                    },
                    usage_details={"input": 38, "output": 22, "total": 60},
                    cost_details=COST_DETAILS,
                )

            trace.update(
                output={
                    "answer": "The bank declined the authorization. Try another card or contact your bank.",
                    "next_action": "retry_payment",
                }
            )
        return getattr(trace, "trace_id", None)


def create_annotation_score(langfuse, trace_id):
    langfuse.create_score(
        trace_id=trace_id,
        name=SCORE_NAME,
        value=SCORE_VALUE,
        data_type="NUMERIC",
        comment=SCORE_COMMENT,
        metadata={
            "source": "ANNOTATION",
            "annotator": "human-reviewer",
        },
    )


def require_trace_id(trace_id, trace_name):
    if not trace_id:
        raise RuntimeError(f"Langfuse SDK did not provide a trace id for {trace_name}")
    return trace_id


def print_summary(checkout_trace_id):
    print("Sent 1 Python Langfuse SDK trace and 1 annotation score:")
    print(f"- checkout support trace ({checkout_trace_id})")
    print(f"- {SCORE_NAME}={SCORE_VALUE} on checkout support trace")
    print(f"- generation cost_details total: {COST_DETAILS['total']}")
    print("")
    print("Inspect checkout trace cost, observations, and scores:")
    print("agentpond sync")
    print(f"agentpond traces get {checkout_trace_id}")
    print(f"agentpond observations list --traceId {checkout_trace_id}")
    print(f"agentpond scores list --traceId {checkout_trace_id}")


def main():
    require_langfuse_env()
    langfuse = get_client()
    checkout_trace_id = require_trace_id(answer_checkout_question(langfuse), "checkout support trace")
    create_annotation_score(langfuse, checkout_trace_id)
    langfuse.flush()
    print_summary(checkout_trace_id)


if __name__ == "__main__":
    main()

File: examples/typescript/send-traces.ts
import { LangfuseClient } from "@langfuse/client";
import { LangfuseSpanProcessor } from "@langfuse/otel";
import {
	getActiveTraceId,
	propagateAttributes,
	startActiveObservation,
} from "@langfuse/tracing";
import { NodeSDK } from "@opentelemetry/sdk-node";

const SCORE_NAME = "human-quality";
const SCORE_VALUE = 1;
const SCORE_COMMENT = "Human review: answer is accurate and actionable.";
const COST_DETAILS = { input: 0.038, output: 0.044, total: 0.082 };

function requireLangfuseEnv() {
	const missing = [
		"LANGFUSE_BASE_URL",
		"LANGFUSE_PUBLIC_KEY",
		"LANGFUSE_SECRET_KEY",
	].filter((key) => !process.env[key]);
	if (missing.length > 0) {
		throw new Error(
			`Missing required Langfuse environment variables: ${missing.join(", ")}. ` +
				"Set LANGFUSE_BASE_URL, LANGFUSE_PUBLIC_KEY, and LANGFUSE_SECRET_KEY before running this example.",
		);
	}
}

async function answerCheckoutQuestion() {
	let traceId: string | undefined;
	await startActiveObservation("checkout support trace", async (trace) => {
		traceId = getActiveTraceId();
		trace.update({
			input: {
				question: "Why was my card declined?",
				locale: "en-US",
				cart_total: 128.45,
			},
			metadata: {
				example: "agentpond-typescript",
				route: "/support/checkout",
				tenant: "acme-retail",
			},
		});

		await propagateAttributes(
			{
				userId: "user_42",
				sessionId: "sdk-example-session",
				traceName: "checkout support trace",
				metadata: {
					example: "agentpond-typescript",
					feature: "checkout-support",
					release: "2026-06-15",
				},
			},
			async () => {
				await startActiveObservation(
					"load customer order context",
					async (span) => {
						span.update({
							input: { user_id: "user_42", order_id: "ord_2026_001" },
							output: {
								payment_status: "declined",
								decline_code: "insufficient_funds",
								attempts: 1,
							},
						});
					},
				);

				await startActiveObservation(
					"draft support response",
					async (generation) => {
						generation.update({
							model: "gpt-5.5-mini",
							input: [
								{
									role: "system",
									content:
										"Answer checkout support questions with concise next steps.",
								},
								{
									role: "user",
									content: "Why was my card declined?",
								},
							],
							output: {
								answer:
									"The bank declined the authorization. Try another card or contact your bank.",
								confidence: 0.92,
							},
							metadata: { provider: "example-fixture", temperature: 0.2 },
							usageDetails: { input: 38, output: 22, total: 60 },
							costDetails: COST_DETAILS,
						});
					},
					{ asType: "generation" },
				);
			},
		);

		trace.update({
			output: {
				answer:
					"The bank declined the authorization. Try another card or contact your bank.",
				next_action: "retry_payment",
			},
		});
	});
	return traceId;
}

function createAnnotationScore(langfuse: LangfuseClient, traceId: string) {
	langfuse.score.create({
		traceId,
		name: SCORE_NAME,
		value: SCORE_VALUE,
		dataType: "NUMERIC",
		comment: SCORE_COMMENT,
		metadata: {
			source: "ANNOTATION",
			annotator: "human-reviewer",
		},
	});
}

function requireTraceId(
	traceId: string | undefined,
	traceName: string,
): string {
	if (!traceId)
		throw new Error(`Langfuse SDK did not provide a trace id for ${traceName}`);
	return traceId;
}

function printSummary(checkoutTraceId: string) {
	console.log("Sent 1 TypeScript Langfuse SDK trace and 1 annotation score:");
	console.log(`- checkout support trace (${checkoutTraceId})`);
	console.log(`- ${SCORE_NAME}=${SCORE_VALUE} on checkout support trace`);
	console.log(`- generation costDetails total: ${COST_DETAILS.total}`);
	console.log("");
	console.log("Inspect checkout trace cost, observations, and scores:");
	console.log("agentpond sync");
	console.log(`agentpond traces get ${checkoutTraceId}`);
	console.log(`agentpond observations list --traceId ${checkoutTraceId}`);
	console.log(`agentpond scores list --traceId ${checkoutTraceId}`);
}

async function main() {
	requireLangfuseEnv();

	const sdk = new NodeSDK({
		spanProcessors: [new LangfuseSpanProcessor()],
	});
	const langfuse = new LangfuseClient();

	sdk.start();
	try {
		const checkoutTraceId = requireTraceId(
			await answerCheckoutQuestion(),
			"checkout support trace",
		);
		createAnnotationScore(langfuse, checkoutTraceId);
		await langfuse.flush();
		printSummary(checkoutTraceId);
	} finally {
		await sdk.shutdown();
		await langfuse.shutdown();
	}
}

main().catch((error) => {
	console.error(error);
	process.exitCode = 1;
});
